### PR TITLE
New program: angular assignment validation

### PR DESCRIPTION
**Pull request by jeymendez**

### CODE REVIEW

**Review by cossorzano — DISMISSED** [review]

File: xmipp3/tests/test_protocol_angular_graph_consistency.py [hunk 1]
```diff
@@ -0,0 +1,99 @@
...
+
+        protValidate = self.newProtocol(XmippProtAngularGraphConsistency,
+                                objLabel='angular graph consistency',
+                                angularSampling=3,
+                                maximumTargetResolution=10,
+                                numberOfMpi=4, numberOfThreads=1)
+        protValidate.inputParticles.set(protImportParts.outputParticles)
+        protValidate.inputVolumes.set(protImportVols.outputVolume)
+
+        self.launchProtocol(protValidate)
```
Comment (albertmena) [inline xmipp3/tests/test_protocol_angular_graph_consistency.py:98]: The test fails to me:
```
 mpirun -np 4  `which xmipp_mpi_transform_mask` -i Runs/000210_XmippProtAngularGraphConsistency/extra/Iter001/globalAssignment/images.stk --mask circular -30
Protocol failed: Command ' mpirun -np 4  `which xmipp_mpi_transform_mask` -i Runs/000210_XmippProtAngularGraphConsistency/extra/Iter001/globalAssignment/images.stk --mask circular -30' returned non-zero exit status 127.
FAILED: globalAssignment, step 3, time 2022-01-26 12:11:02.787882

```

**Review by albertmena — CHANGES_REQUESTED** [review]

File: xmipp3/protocols/protocol_angular_graph_consistency.py [hunk 2]
```diff
@@ -0,0 +1,588 @@
...
+# *
+# **************************************************************************
+
+from glob import glob
+import math
+from os.path import join, isfile, exists
+from shutil import copyfile, copy
+import os
+
+#from pyworkflow.object import Float, String
```
[the remark this answers is not in the document]
Comment (jeymendez) [reply xmipp3/protocols/protocol_angular_graph_consistency.py:33]: done

File: xmipp3/protocols/protocol_angular_graph_consistency.py [hunk 3]
```diff
@@ -0,0 +1,588 @@
...
+        form.addParallelSection(threads=1, mpi=1)
+
+    # --------------------------- INSERT steps functions --------------------------------------------
+
+    def _insertAllSteps(self):
+        self.imgsFn = self._getExtraPath('images.xmd')
+        convertId = self._insertFunctionStep('convertInputStep',
+                                             self.inputParticles.get().getObjId())
+        self.TsOrig = self.inputParticles.get().getSamplingRate()
+        sym = self.symmetryGroup.get()
```
[the remark this answers is not in the document]
Comment (jeymendez) [reply xmipp3/protocols/protocol_angular_graph_consistency.py:116]: done

File: xmipp3/protocols/protocol_angular_graph_consistency.py [hunk 4]
```diff
@@ -0,0 +1,588 @@
...
+        
+        img = ImageHandler()
+        if isinstance(self.inputVolumes.get(), SetOfVolumes):
+            i = 1
+            for vol in self.inputVolumes.get():
+                fnVol = join(fnDirCurrent, "volume.vol") 
+                img.convert(vol, fnVol)
+                if newXdim != vol.getDim()[0]:
+                    self.runJob('xmipp_transform_window', "-i %s --size %d" % (fnVol, newXdim), numberOfMpi=1)
+                i += 1
```
[the remark this answers is not in the document]
Comment (jeymendez) [reply xmipp3/protocols/protocol_angular_graph_consistency.py:261]: done

File: xmipp3/protocols/protocol_angular_graph_consistency.py [hunk 5]
```diff
@@ -0,0 +1,588 @@
...
+            cleanPath(join(fnDirAssignment, "gallery_angles%s.doc" % subset))
+            moveFile(join(fnDirAssignment, "gallery%s.doc" % subset), fnGalleryMd)
+            fnAngles = join(fnGlobal, "anglesDisc%s.xmd" % subset)
+            for j in range(1, numberGroups + 1):
+                fnAnglesGroup = join(fnDirAssignment, "angles_group%03d%s.xmd" % (j, subset))
+                if not exists(fnAnglesGroup):
+                    if ctfPresent:
+                        fnGroup = "ctfGroup%06d@%s/ctf_groups.xmd" % (j, fnDirAssignment)
+                        fnCTF = "%d@%s/ctf_ctf.stk" % (j, fnDirAssignment)
+                        fnGalleryGroup = fnGallery = join(fnDirAssignment, "gallery%s_%06d.stk" % (subset, j))
```
[the remark this answers is not in the document]
Comment (jeymendez) [reply xmipp3/protocols/protocol_angular_graph_consistency.py:338]: done

File: xmipp3/protocols/protocol_angular_graph_consistency.py [hunk 6]
```diff
@@ -0,0 +1,588 @@
...
+            newXdim = int(40)
+            TsCurrent = Xdim * (self.TsOrig / newXdim)
+        elif newXdim % 2 == 1:
+            newXdim += 1
+            TsCurrent = Xdim * (self.TsOrig / newXdim)
+        self.writeInfoField(fnDir, "sampling", emlib.MDL_SAMPLINGRATE, TsCurrent)
+        self.writeInfoField(fnDir, "size", emlib.MDL_XSIZE, newXdim)
+        
+        # Prepare particles
+        fnDir0 = self._getExtraPath("Iter000")
```
[the remark this answers is not in the document]
Comment (jeymendez) [reply xmipp3/protocols/protocol_angular_graph_consistency.py:460]: done

File: xmipp3/protocols/protocol_angular_graph_consistency.py [hunk 7]
```diff
@@ -0,0 +1,588 @@
...
+            moveFile("%s/ctf_images.sel" % fnDirAssignment, "%s/ctf_groups.xmd" % fnDirAssignment)
+            cleanPath("%s/ctf_split.doc" % fnDirAssignment)
+            mdInfo = emlib.MetaData("numberGroups@%s" % join(fnDirAssignment, "ctfInfo.xmd"))
+            fnCTFs = "%s/ctf_ctf.stk" % fnDirAssignment
+            numberGroups = mdInfo.getValue(emlib.MDL_COUNT, mdInfo.firstObject())
+            ctfPresent = True
+        else:
+            numberGroups = 1
+            ctfPresent = False
+            fnCTFs = ""
```
[the remark this answers is not in the document]
Comment (jeymendez) [reply xmipp3/protocols/protocol_angular_graph_consistency.py:318]: done

File: xmipp3/protocols/protocol_angular_graph_consistency.py [hunk 8]
```diff
@@ -0,0 +1,588 @@
...
+                        fnGroup = "ctfGroup%06d@%s/ctf_groups.xmd" % (j, fnDirAssignment)
+                        fnCTF = "%d@%s/ctf_ctf.stk" % (j, fnDirAssignment)
+                        fnGalleryGroup = fnGallery = join(fnDirAssignment, "gallery%s_%06d.stk" % (subset, j))
+                        fnGalleryGroupMd = fnGallery = join(fnDirAssignment, "gallery%s_%06d.xmd" % (subset, j))
+                        self.runJob("xmipp_transform_filter", "-i %s -o %s --fourier binary_file %s --save_metadata_stack %s --keep_input_columns" % \
+                                    (fnGalleryMd, fnGalleryGroup, fnCTF, fnGalleryGroupMd),
+                                    numberOfMpi=min(self.numberOfMpi.get(), 24))
+                    else:
+                        fnGroup = fnImgs
+                        fnGalleryGroup = fnGallery
```
[the remark this answers is not in the document]
Comment (jeymendez) [reply xmipp3/protocols/protocol_angular_graph_consistency.py:345]: done

File: xmipp3/protocols/protocol_angular_graph_consistency.py [hunk 9]
```diff
@@ -0,0 +1,588 @@
...
+        if self.checkInfoField(fnDir, "count"):
+            state = self.readInfoField(fnDir, "count", emlib.MDL_COUNT)
+            if state >= 2:
+                return
+
+        print("Preparing references to sampling rate=", TsCurrent)
+        fnMask = ''
+        newXdim = self.readInfoField(fnDir, "size", emlib.MDL_XSIZE)
+        oldXdim = self.readInfoField(fnDirPrevious, "size", emlib.MDL_XSIZE)
+        i=1
```
[the remark this answers is not in the document]
Comment (jeymendez) [reply xmipp3/protocols/protocol_angular_graph_consistency.py:491]: done

**Comment by DStrelak** [comment]
@jeymendez , please click on `Resolve conversation` or simply change the part of the code related to review (which will mark it as outdated). I'm getting way too many emails :D
Also, please first fix your code so that it passes the SonarCloud Analysis, and includes at least basic tests of the functionality. After that, we can do a second round of the reviews.
Feel free to contact either me or @albertmena  if you need to discuss the code in details or need help with those issues.

File: xmipp3/tests/test_protocol_angular_graph_consistency.py (hunk 1 above)
Comment (jeymendez) [reply xmipp3/tests/test_protocol_angular_graph_consistency.py]: Hi @albertmena,
in my my laptop the test pass correctly. 
I have been trying to install again scipion3 in hertz-dos, in order to try to reproduce the error but i have problems installing it there.
command gcc --version in hertz-dos shows gcc 4 installed... I have compiled and installed in my home folder gcc-10.2 and gcc-8.4 and still having issues compiling xmipp.

Comment (DStrelak) [reply xmipp3/tests/test_protocol_angular_graph_consistency.py]: For me the test passed. 
But it took an hour, which is... not very good.
I'll run some basic profiling to let you know where to have a look

Comment (DStrelak) [reply xmipp3/tests/test_protocol_angular_graph_consistency.py]: So, you spend majority of your time in the `ccMatrix` by `recomputePlanR2C()`.
Get rid of the `CorrelationAux`on line 760 (you don't use i for other things but trosmformer1 anyway), and use some transformer that you store as the class attribute. 
Do not use it for other work but for IFFT.
Check the sizes of your images, consider cropping them to even sizes (i.e. sizex % 2 == 0 && sizey % 2 == 0), as that typically leads to faster processsing.
Like that, you should avoid the penalization of replanning.
Other option is to use `aft.h`, which provides both CPU and CUDA version of FFT.

Comment (albertmena) [reply xmipp3/tests/test_protocol_angular_graph_consistency.py]: now the test pass for me!

**Review by albertmena — APPROVED** [review]

**Pull request merged**